text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+types. However, Rust does not provide stability guarantees for its niche
+optimizations, so crABI cannot simultaneously have a documented ABI (defining
+an exact set of specific niche optimizations and no others) and maintain layout
+compatibility with Rust: Rust could not then do any further niche optimizations
+not already specified by crABI, even in an edition (because types and their
+layouts are intercompatible between editions). Thus, this crABI proposal
+(reluctantly) specifies *separate* `Option`, `Result`, and tuple types, to
+avoid this compatibility hazard.
+
+Making a structure compatible with crABI requires declaring it with either
+`repr(crabi)` or `repr(C)`. This makes it difficult to adapt a type from
+another library, or to add a crABI interface to an existing Rust library while
+preserving its data types.
+
+crABI could (instead of or in addition to `repr(crabi)`) provide a mechanism
+that "transforms" a type into the `repr(crabi)` equivalent of that type, to avoid
+having to define a separate type. However, this would not handle the problem of
+translating between the types, which in the general case of a "deep"
+translation could require copying arbitrary in-memory data structures. crABI
+does not aim to solve that problem.
+
+The translation of slices and similar uses structs containing pointer/length
+pairs, rather than inlining the pointer and length as separate arguments.
+[As noted above][types], this is typically passed and returned in an efficient
+fashion on major targets. However, in some languages, such as C, this will
+require separately defining a structure and then using that structure. This
+still seems preferable, though, as combining the two into one struct allows for
+uniform handling between arguments, return values, and fields, as well as
+keeping the pointer and length more strongly associated.
+
+@programmerjake made a
+[proposal](https://github.com/rust-lang/rfcs/pull/3470#issuecomment-1674249638)
+([sample usage](https://github.com/rust-lang/rfcs/pull/3470#issuecomment-1674265515))
+to modify the standard impl of `Drop` for `Box` to allow plugging in an
+arbitrary function (via a `BoxDrop` trait), to drop the `Box` as a whole. This
+would be generally useful (e.g. for object pooling), and would then permit
+crABI to define a `box_drop` function that calls an FFI function to free the
+object. If we accepted that proposal, it would make sense to use it to
+represent crABI boxes.
+
+# Prior art
+[prior-art]: #prior-art
+
+Some potential sources of inspiration:
+
+- WebAssembly Interface Types
+- Swift's stable ABI
+- The `abi_stable` crate (which aims for Rust-to-Rust stability, not
+  cross-language interoperation, but it still serves as a useful reference)
+- `stabby`
+- UniFFI
+- Diplomat
+- C++'s various ABIs (and the history of its ABI changes). crABI does not,
+  however, aim for compatibility with or supersetting of any particular C++
+  ABI.
+- Many, many interface description languages (IDLs).
```

Comment:
The wasm component model has a list type, but currently only allowes owned values. C++ has `string_view` as `&str` counterpart and `span<T>` as `&[T]` counterpart.

Reply:
Go does: https://go.googlesource.com/go/+/refs/heads/dev.regabi/src/cmd/compile/internal-abi.md

Reply:
> The slice type `[]T` is a sequence of a `*[cap]T` pointer to the slice backing store, an int giving the len of the slice, and an `int` giving the cap of the slice.

That sounds like it may be storage-compatible with the `struct slice { uint8_t *data; size_t len; }` defined here, which is kind of nice. It seems like C compatibility isn't necessarily a goal of go's ABI though so that might be it.

Reply:
I guess you meant (size, pointer), but while libstdc++ uses that one, libc++ (i.e. LLVM's) and Microsoft's use (pointer, size) instead.

For `span` (dynamic), all three use (pointer, size).

Well, at least the versions I looked at.